openscenario/openscenario_interpreter/src/syntax/relative_clearance_condition.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright 2015 TIER IV, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <openscenario_interpreter/reader/attribute.hpp>
+#include <openscenario_interpreter/reader/element.hpp>
+#include <openscenario_interpreter/syntax/entities.hpp>
+#include <openscenario_interpreter/syntax/relative_clearance_condition.hpp>
+#include <openscenario_interpreter/utility/print.hpp>
+#include <unordered_map>
+
+namespace openscenario_interpreter
+{
+inline namespace syntax
+{
+RelativeClearanceCondition::RelativeClearanceCondition(
+  const pugi::xml_node & node, Scope & scope, const TriggeringEntities & triggering_entities)
+: Scope(scope),
+  distance_backward(readAttribute<Double>("distanceBackward", node, scope, 0.0)),
+  distance_forward(readAttribute<Double>("distanceForward", node, scope, 0.0)),
+  free_space(readAttribute<Boolean>("freeSpace", node, scope)),
+  opposite_lanes(readAttribute<Boolean>("oppositeLanes", node, scope)),
+  relative_lane_range(readElements<RelativeLaneRange, 0>("RelativeLaneRange", node, scope)),
+  entity_refs([&]() {
+    auto entities = readElements<EntityRef, 0>("EntityRef", node, scope);
+    if (entities.empty()) {
+      for (const auto & entity : *global().entities) {
+        if (
+          std::find_if(
+            triggering_entities.entity_refs.begin(), triggering_entities.entity_refs.end(),
+            [&](const auto & triggering_entity) {
+              return triggering_entity.name() == entity.first;
+            }) == triggering_entities.entity_refs.end()) {
+          entities.emplace_back(entity.first);
+        }
+      }
+    }
+    return entities;
+  }()),
```

Comment:
Why is the initialization of entity_refs implemented like this? It seems that the behavior is programmed differently from the conventions in the interpreter and from the OpenSCENARIO XML standard.

Also, since EntitySelection is supported, the name referenced by EntityRef may be an alias that points to multiple entities, so each syntax should be stored in Entity type instead of EntityRef. For more information, please see the following pull request: https://github.com/tier4/scenario_simulator_v2/pull/1059

Reply:
> Why is the initialization of entity_refs implemented like this? It seems that the behavior is programmed differently from the conventions in the interpreter and from the OpenSCENARIO XML standard.

What is exactly "programmed differently from the conventions in the interpreter and from the OpenSCENARIO XML standard"?
I didn't get it, so there may be problems with my understanding of the conventions and starndards.
Could you tell me your expectations and the difference of current code?

Note: Lines from 36 to 47 are implemented for the standard description `If it is not used then all entities are considered.` for `entityRef`.
Here, this specification is performed at the same time as initialization and written to the `entity_refs` member.
Is what is being pointed out here that the problem is that what is actually written in the scenario and the result of class construction are different?

> Also, since EntitySelection is supported, the name referenced by EntityRef may be an alias that points to multiple entities, so each syntax should be stored in Entity type instead of EntityRef. For more information, please see the following pull request: https://github.com/tier4/scenario_simulator_v2/pull/1059

OK. I'll fix.

Reply:
> Note: Lines from 36 to 47 are implemented for the standard description If it is not used then all entities are considered. for entityRef.

If you want to consider all entities, you should select entities that are `is_added` out of global()->entities. The code you wrote is "global()->entities whose names are contained in triggering_entities.entitiy_refs". So my question is what is the need to touch triggering_entities.entity_refs to search for "all entities".

Reply:
> you should select entities that are is_added out of global()->entities

Sorry, I didn't know this was necessary to filter with `is_added`. I'll try adding a filter by `is_added`.

> The code you wrote is "global()->entities whose names are contained in triggering_entities.entitiy_refs".

This is the opposite.
Note that the operator used in the comparison between `find_if` and `end` is "==", not "!=".
(The fact that you made a mistake here might indicate that my code is difficult to read.)

> Also, the number of "all entities" varies since the construct of this class, considering the existence of AddEntityAction and DeleteEntityAction. Therefore, I think that the search for "all entities" should be performed each time a condition is determined.

Oh, I missed it. Thank you for pointing out my missed cases!

Reply:
> This is the opposite.
Note that the operator used in the comparison between find_if and end is "==", not "!=".
(The fact that you made a mistake here might indicate that my code is difficult to read.)

I was wrong about this, sorry.